src/path.c

```diff
@@ -1,90 +1,114 @@
+	    {"LLF", MOVE_LEFT_180}, {"RRF", MOVE_RIGHT_180},
+	    {"", MOVE_NONE},
+	},
+	.from_diagonal = (struct translation[]){
+	    {"LL", MOVE_LEFT}, {"RR", MOVE_RIGHT},
+	    {"LR", MOVE_LEFT}, {"RL", MOVE_RIGHT},
+	    {"LF", MOVE_LEFT}, {"RF", MOVE_RIGHT},
+	    {"", MOVE_NONE},
+	},
+    },
+    [PATH_DIAGONALS] = {
+	.from_straight = (struct translation[]){
+	    {"LF", MOVE_LEFT_90},      {"RF", MOVE_RIGHT_90},
+	    {"LR", MOVE_LEFT_TO_45},   {"RL", MOVE_RIGHT_TO_45},
+	    {"LLF", MOVE_LEFT_180},    {"RRF", MOVE_RIGHT_180},
+	    {"LLR", MOVE_LEFT_TO_135}, {"RRL", MOVE_RIGHT_TO_135},
+	    {"", MOVE_NONE},
+	},
+	.from_diagonal = (struct translation[]){
+	    {"LR", MOVE_DIAGONAL},       {"RL", MOVE_DIAGONAL},
+	    {"LF", MOVE_LEFT_FROM_45},   {"RF", MOVE_RIGHT_FROM_45},
+	    {"LLR", MOVE_LEFT_DIAGONAL}, {"RRL", MOVE_RIGHT_DIAGONAL},
+	    {"LLF", MOVE_LEFT_FROM_135}, {"RRF", MOVE_RIGHT_FROM_135},
+	    {"", MOVE_NONE},
+	},
+    },
+};
+// clang-format on
 
 /**
- * @brief Translate the source path when coming from a diagonal movement.
+ * @brief Translate the source path to a smooth path.
  *
- * @return Number of movements translated from source.
+ * @param[in] source Raw path to translate from.
+ * @param[in] path_language Language to use for the translation.
+ * @param[in] from_straight If we come from a straight movement.
  */
-static int from_diagonal(void)
+static struct translation translate(char *source, enum path_language language,
+				    bool from_straight)
 {
-	if (!strncmp(source, "B", 1)) {
-		*destination = MOVE_START;
-		return 1;
-	}
-	if (!strncmp(source, "L", 1)) {
-		*destination = MOVE_LEFT;
-		return 1;
-	}
-	if (!strncmp(source, "R", 1)) {
-		*destination = MOVE_RIGHT;
-		return 1;
+	int length;
+	struct translation *candidate;
+
+	if (from_straight)
+		candidate = languages[language].from_straight;
+	else
+		candidate = languages[language].from_diagonal;
+	while (true) {
+		length = strlen(candidate->from);
+		if (!length)
+			break;
+		if (!strncmp(source, candidate->from, length))
+			return *candidate;
+		candidate++;
 	}
-	return 0;
+	return *candidate;
 }
 
 /**
  * @brief Make a smooth path out of a raw, exploration path.
  *
- * @param[in] raw_path Raw path to smooth.
- * @param[out] smooth_path Array to write the smooth path to.
+ * @param[in] source Raw path to smooth.
+ * @param[out] destination Array to write the smooth path to.
+ * @param[in] path_language Language to use for the translation.
  */
-void make_smooth_path(char *raw_path, enum movement *smooth_path)
+void make_smooth_path(char *source, enum movement *destination,
+		      enum path_language language)
 {
-	source = raw_path;
-	destination = smooth_path;
+	bool straight = false;
+	struct translation translated;
+
 	while (true) {
 		if (*source == '\0')
 			break;
+		if (*source == 'B') {
+			*destination++ = MOVE_START;
+			source++;
+			continue;
+		}
 		if (*source == 'S') {
 			*destination++ = MOVE_STOP;
```

Comment:
Because I am `break;`ing right after. So no need to keep processing the source path.

Reply:
Then I found that there is no code after the `while` loop, so we can just return...

7c2e4ab74f9a3741ac19d07bcdeb24f07abba632

Added a new commit instead of a fixup because that code is already in master, not anything new added with this "diagonals" implementation. :wink:

Reply:
b50b929d41a5760a8dda7c560263c235a320f971

Reply:
:+1: